staging/src/k8s.io/legacy-cloud-providers/gce/gce_loadbalancer_metrics.go

```diff
@@ -0,0 +1,165 @@
+// +build !providerless
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gce
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+const (
+	label = "feature"
+)
+
+var (
+	metricsInterval = 10 * time.Minute
+	l4ILBCount = metrics.NewGaugeVec(
+		&metrics.GaugeOpts{
+			Name: "number_of_l4_ilbs",
+			Help: "Number of L4 ILBs",
+		},
+		[]string{label},
+	)
+)
+
+// init registers L4 internal loadbalancer usage metrics.
+func init() {
+	klog.V(3).Infof("Registering Service Controller usage metrics %v", l4ILBCount)
+	legacyregistry.MustRegister(l4ILBCount)
+}
+
+// LoadBalancerMetrics is a cache that contains loadbalancer service resource
+// states for computing usage metrics.
+type LoadBalancerMetrics struct {
+	// l4ILBServiceMap is a map of service key and L4 ILB service state.
+	l4ILBServiceMap map[string]L4ILBServiceState
+
+	sync.Mutex
+}
+
+type feature string
+
+func (f feature) String() string {
+	return string(f)
+}
+
+const(
+	l4ILBService      = feature("L4ILBService")
+	l4ILBGlobalAccess = feature("L4ILBGlobalAccess")
+	l4ILBCustomSubnet = feature("L4ILBCustomSubnet")
+)
+
+// L4ILBServiceState contains Internal Loadbalancer feature states as specified
+// in k8s Service.
+type L4ILBServiceState struct {
+	// EnabledGlobalAccess specifies if Global Access is enabled.
+	EnabledGlobalAccess bool
+	// EnabledCustomSubNet specifies if Custom Subnet is enabled.
+	EnabledCustomSubnet bool
```

Comment:
It would be great to have an InError bool to expose a count of ILB services which are in error state. It can be a follow up.

Reply:
Yes, good idea. But we can only monitor the errors that originate from `ensureInternalLoadBalancer` or its caller `ensureLoadBalancer`. 

Would it be okay to count the services those error out in just `ensureInternalLoadBalancer` method?